<!DOCTYPE HTML>
<html lang="en">
<head>
<!-- Generated by javadoc (17) -->
<title>CatchException (catch-exception-project 2.1.3 API)</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="description" content="declaration: package: com.googlecode.catchexception, class: CatchException">
<meta name="generator" content="javadoc/ClassWriterImpl">
<link rel="stylesheet" type="text/css" href="../../../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../../../script-dir/jquery-ui.min.css" title="Style">
<link rel="stylesheet" type="text/css" href="../../../jquery-ui.overrides.css" title="Style">
<script type="text/javascript" src="../../../script.js"></script>
<script type="text/javascript" src="../../../script-dir/jquery-3.5.1.min.js"></script>
<script type="text/javascript" src="../../../script-dir/jquery-ui.min.js"></script>
</head>
<body class="class-declaration-page">
<script type="text/javascript">var evenRowColor = "even-row-color";
var oddRowColor = "odd-row-color";
var tableTab = "table-tab";
var activeTableTab = "active-table-tab";
var pathtoroot = "../../../";
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<div class="flex-box">
<header role="banner" class="flex-header">
<nav role="navigation">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="top-nav" id="navbar-top">
<div class="skip-nav"><a href="#skip-navbar-top" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar-top-firstrow" class="nav-list" title="Navigation">
<li><a href="../../../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="class-use/CatchException.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../index-all.html">Index</a></li>
<li><a href="../../../help-doc.html#class">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor-summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method-summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor-detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method-detail">Method</a></li>
</ul>
</div>
<div class="nav-list-search"><label for="search-input">SEARCH:</label>
<input type="text" id="search-input" value="search" disabled="disabled">
<input type="reset" id="reset-button" value="reset" disabled="disabled">
</div>
</div>
<!-- ========= END OF TOP NAVBAR ========= -->
<span class="skip-nav" id="skip-navbar-top"></span></nav>
</header>
<div class="flex-content">
<main role="main">
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="sub-title"><span class="package-label-in-type">Package</span>&nbsp;<a href="package-summary.html">com.googlecode.catchexception</a></div>
<h1 title="Class CatchException" class="title">Class CatchException</h1>
</div>
<div class="inheritance" title="Inheritance Tree"><a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html" title="class or interface in java.lang" class="external-link">java.lang.Object</a>
<div class="inheritance">com.googlecode.catchexception.CatchException</div>
</div>
<section class="class-description" id="class-description">
<hr>
<div class="type-signature"><span class="modifiers">public class </span><span class="element-name"><a href="../../../src-html/com/googlecode/catchexception/CatchException.html#line-207">CatchException</a></span>
<span class="extends-implements">extends <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html" title="class or interface in java.lang" class="external-link">Object</a></span></div>
<div class="block">The methods of this class catch and verify exceptions in <em>a single line of
 code</em> and make them available for further analysis.

 This Javadoc content is also available on the <a href="http://code.google.com/p/catch-exception/">catch-exception</a> web
 page.

 <strong>Documentation</strong>

 <b> <a href="#1">1. How to use catch-exception?</a>

 <a href="#2">2. What is this stuff actually good for?</a>

 <a href="#3">3. How does it work internally?</a>

 <a href="#4">4. When is the caught exception reset?</a>

 <a href="#5">5. My code throws a ClassCastException. Why?</a>

 <a href="#6">6. The exception is not caught. Why?</a>

 <a href="#7">7. Do I have to care about memory leaks?</a>

 <a href="#8">8. The caught exception is not available in another thread.
 Why?</a>

 <a href="#9">9. How do I catch an exception thrown by a static method?</a>

 <a href="#11">11. Can I catch errors instead of exceptions?</a>




 </b>

 <strong>1. How to use catch-exception?</strong>

 The most basic usage is:
 <code>import static com.googlecode.catchexception.CatchException.*;

 // call customerService.prepareBilling(Prize.Zero)
 // and catch the exception if any is thrown
 catchException(customerService).prepareBilling(Prize.Zero);

 // assert that an IllegalArgumentException was thrown
 assert caughtException() instanceof IllegalArgumentException;</code>

 You can combine the two lines of code in a single one if you like:
 <code>// call customerService.prepareBilling(Prize.Zero)
 // and throw an ExceptionNotThrownAssertionError if
 // the expected exception is not thrown
 verifyException(customerService, IllegalArgumentException.class).prepareBilling(Prize.Zero);</code>
 There is a minor difference between both variants. In the first variant you
 must start the JVM with option <code>-ea</code> to enable the assertion. The
 second variant does not use JDK assertions and ,therefore, always verifies
 the caught exception.

 A third variant allows you to select the type of exceptions you want to catch
 (no verification involved):
 <code>// catch IllegalArgumentExceptions but no other exceptions
 catchException(customerService, IllegalArgumentException.class).prepareBilling(Prize.Zero);</code>

 The fourth and last variant verifies that some exception is thrown, i.e. the
 type of the exception does not matter:
 <code>verifyException(customerService).prepareBilling(Prize.Zero);</code>

 In all variants you can use <code>caughtException()</code> afterwards to
 inspect the caught exception.

 Finally, there some alternative ways to catch and verify exceptions:
 <ul>
 <li><a href="apis/BDDCatchException.html" title="class in com.googlecode.catchexception.apis"><code>BDDCatchException</code></a> - a BDD-like approach,
 <li> <a href="apis/CatchExceptionHamcrestMatchers.html" title="class in com.googlecode.catchexception.apis"><code>CatchExceptionHamcrestMatchers</code></a> - Hamcrest assertions
 </ul>
 <strong>2. What is this stuff actually good for?</strong>

 This class targets concise and robust code in tests. Dadid Saff, a commiter
 to JUnit, has <a href="http://shareandenjoy.saff.net/2006/12/assertthrownexception_20.html">discussed</a> this approach in 2007. Let me summarize the arguments here.

 There are two advantages of the approach proposed here in comparison to the
 use of try/catch blocks.
 <ul>
 <li>The test is more concise and easier to read.
 <li>The test cannot be corrupted by a missing assertion. Assume you forgot to
 type <code>fail()</code> behind the method call that is expected to throw an
 exception.
 </ul>

 There are also some advantages of this approach in comparison to test
 runner-specific mechanisms that catch and verify exceptions.
 <ul>
 <li>A single test can verify more than one thrown exception.
 <li>The test can verify the properties of the thrown exception after the
 exception is caught.
 <li>The test can specify by which method call the exception must be thrown.
 <li>The test does not depend on a specific test runner (JUnit4, TestNG).
 </ul>

 <strong>3. How does it work internally?</strong>

 The method <code>catchException(obj)</code> wraps the given object with a
 proxy that catches the exception, then (optionally) verifies the exception,
 and finally attaches the exception to the current <a id="threadlocal">thread</a> for further analysis. The <a href="#proxies">known limitations</a> for proxies apply.

 Is both memory consumption and runtime a concern for you? Then use try/catch
 blocks instead of this class. Because in this case the creation of proxies is
 an unnecessary overhead. If only either memory consumption or runtime is an
 issue for you, feel free to configure the cache of the underlying proxy
 factories as appropriate.
 <strong>4. When is the caught exception reset?</strong>

 The Method <a href="#caughtException()"><code>caughtException()</code></a> returns the exception thrown by the
 last method call on a proxied object in the current thread, i.e. it is reset
 by calling a method on the proxied object. If the called method has not
 thrown an exception, <code>caughtException()</code> returns null.

 To reset the caught exception manually, call <a href="#resetCaughtException()"><code>resetCaughtException()</code></a>.
 At the moment there is no way to reset exceptions that have been caught in
 other threads.
 <strong>5. My code throws a ClassCastException. Why?</strong>

 Example:
 <code>StringBuilder sb = new StringBuilder();
 catchException(sb).charAt(-2); // throws ClassCastException</code>

 Probably you have tested a final class. Proxy factories usually try to
 subclass the type of the proxied object. This is not possible if the original
 class is final. But there is a way out. If the tested method belongs to an
 interface, then you can cast the argument (here: <code>sb</code>) to that
 interface or ,easier, change the declared type of the argument to the
 interface type. This works because the created proxy is not longer required
 to have the same type as the original class but it must only have the same
 interface.
 <code>// first variant
 StringBuilder sb = new StringBuilder();
 catchException((CharSequence) sb).charAt(-2); // works fine

 // second variant
 CharSequence sb = new StringBuilder();
 catchException(sb).charAt(-2); // works fine</code> If the tested
 method does no belong to an interface fall back to the try/catch-blocks or
 use <a href="http://code.google.com/p/catch-exception/wiki/Dependencies">Powermock
 </a>.
 <code>// example for
 PowerMock with JUnit4
 &#064;RunWith(PowerMockRunner.class)
 &#064;PrepareForTest({ MyFinalType.class })
 public class MyTest {
 </code>
 <strong>6. Do I have to care about memory leaks?</strong>

 This library uses a <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/ThreadLocal.html" title="class or interface in java.lang" class="external-link"><code>ThreadLocal</code></a>. ThreadLocals are known to cause
 memory leaks if they refer to a class the garbage collector would like to
 collect. If you use this library only for testing, then memory leaks do not
 worry you. If you use this library for other purposes than testing, you
 should care.

 <strong>7. The caught exception is not available in another thread. Why?</strong>

 The caught exception is saved <a href="#threadlocal">at the thread</a> the
 exception is thrown in. This is the reason the exception is not visible
 within any other thread.
 <strong>8. Is there a way to get rid of the throws clause in my test
 method?</strong>

 Example:
 <code>public void testSomething() throws Exception {
 ...
 catchException(obj).do(); // do() throws a checked exception</code> No,
 although the exception is always caught you cannot omit the throws clause in
 your test method.
 <strong>11. Can I catch errors instead of exceptions?</strong>

 Yes, have a look at
 <code>com.googlecode.catchexception.throwable.CatchThrowable</code> (in module
 catch-throwable).</div>
<dl class="notes">
<dt>Since:</dt>
<dd>16.09.2011</dd>
<dt>Author:</dt>
<dd>rwoo</dd>
</dl>
</section>
<section class="summary">
<ul class="summary-list">
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<li>
<section class="constructor-summary" id="constructor-summary">
<h2>Constructor Summary</h2>
<div class="caption"><span>Constructors</span></div>
<div class="summary-table two-column-summary">
<div class="table-header col-first">Constructor</div>
<div class="table-header col-last">Description</div>
<div class="col-constructor-name even-row-color"><code><a href="#%3Cinit%3E()" class="member-name-link">CatchException</a>()</code></div>
<div class="col-last even-row-color">&nbsp;</div>
</div>
</section>
</li>
<!-- ========== METHOD SUMMARY =========== -->
<li>
<section class="method-summary" id="method-summary">
<h2>Method Summary</h2>
<div id="method-summary-table">
<div class="table-tabs" role="tablist" aria-orientation="horizontal"><button id="method-summary-table-tab0" role="tab" aria-selected="true" aria-controls="method-summary-table.tabpanel" tabindex="0" onkeydown="switchTab(event)" onclick="show('method-summary-table', 'method-summary-table', 3)" class="active-table-tab">All Methods</button><button id="method-summary-table-tab1" role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" onclick="show('method-summary-table', 'method-summary-table-tab1', 3)" class="table-tab">Static Methods</button><button id="method-summary-table-tab4" role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" onclick="show('method-summary-table', 'method-summary-table-tab4', 3)" class="table-tab">Concrete Methods</button></div>
<div id="method-summary-table.tabpanel" role="tabpanel">
<div class="summary-table three-column-summary" aria-labelledby="method-summary-table-tab0">
<div class="table-header col-first">Modifier and Type</div>
<div class="table-header col-second">Method</div>
<div class="table-header col-last">Description</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code>static void</code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code><a href="#catchException(com.googlecode.catchexception.ThrowingCallable)" class="member-name-link">catchException</a><wbr>(<a href="ThrowingCallable.html" title="interface in com.googlecode.catchexception">ThrowingCallable</a>&nbsp;actor)</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4">
<div class="block">Use it to catch an exception and to get access to the thrown exception
 (for further verifications).</div>
</div>
<div class="col-first odd-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code>static void</code></div>
<div class="col-second odd-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code><a href="#catchException(com.googlecode.catchexception.ThrowingCallable,java.lang.Class)" class="member-name-link">catchException</a><wbr>(<a href="ThrowingCallable.html" title="interface in com.googlecode.catchexception">ThrowingCallable</a>&nbsp;actor,
 <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html" title="class or interface in java.lang" class="external-link">Class</a>&lt;? extends <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html" title="class or interface in java.lang" class="external-link">Exception</a>&gt;&nbsp;clazz)</code></div>
<div class="col-last odd-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4">
<div class="block">Use it to catch an exception of a specific type and to get access to the
 thrown exception (for further verifications).</div>
</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code>private static void</code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code><a href="#catchException(com.googlecode.catchexception.ThrowingCallable,java.lang.Class,boolean)" class="member-name-link">catchException</a><wbr>(<a href="ThrowingCallable.html" title="interface in com.googlecode.catchexception">ThrowingCallable</a>&nbsp;actor,
 <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html" title="class or interface in java.lang" class="external-link">Class</a>&lt;? extends <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html" title="class or interface in java.lang" class="external-link">Exception</a>&gt;&nbsp;clazz,
 boolean&nbsp;assertException)</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4">&nbsp;</div>
<div class="col-first odd-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code>static &lt;E extends <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html" title="class or interface in java.lang" class="external-link">Exception</a>&gt;<br>E</code></div>
<div class="col-second odd-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code><a href="#caughtException()" class="member-name-link">caughtException</a>()</code></div>
<div class="col-last odd-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4">
<div class="block">Returns the exception caught during the last call in the current thread.</div>
</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code>static &lt;E extends <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html" title="class or interface in java.lang" class="external-link">Exception</a>&gt;<br>E</code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code><a href="#caughtException(java.lang.Class)" class="member-name-link">caughtException</a><wbr>(<a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html" title="class or interface in java.lang" class="external-link">Class</a>&lt;E&gt;&nbsp;caughtExceptionType)</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4">&nbsp;</div>
<div class="col-first odd-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code>static void</code></div>
<div class="col-second odd-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code><a href="#resetCaughtException()" class="member-name-link">resetCaughtException</a>()</code></div>
<div class="col-last odd-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4">
<div class="block">Sets the <a href="#caughtException()"><code>caught exception</code></a> to null.</div>
</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code>private static void</code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code><a href="#validateArguments(com.googlecode.catchexception.ThrowingCallable,java.lang.Class)" class="member-name-link">validateArguments</a><wbr>(<a href="ThrowingCallable.html" title="interface in com.googlecode.catchexception">ThrowingCallable</a>&nbsp;actor,
 <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html" title="class or interface in java.lang" class="external-link">Class</a>&lt;? extends <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html" title="class or interface in java.lang" class="external-link">Exception</a>&gt;&nbsp;clazz)</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4">&nbsp;</div>
<div class="col-first odd-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code>static void</code></div>
<div class="col-second odd-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code><a href="#verifyException(com.googlecode.catchexception.ThrowingCallable)" class="member-name-link">verifyException</a><wbr>(<a href="ThrowingCallable.html" title="interface in com.googlecode.catchexception">ThrowingCallable</a>&nbsp;actor)</code></div>
<div class="col-last odd-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4">
<div class="block">Use it to verify that an exception is thrown and to get access to the
 thrown exception (for further verifications).</div>
</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code>static void</code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4"><code><a href="#verifyException(com.googlecode.catchexception.ThrowingCallable,java.lang.Class)" class="member-name-link">verifyException</a><wbr>(<a href="ThrowingCallable.html" title="interface in com.googlecode.catchexception">ThrowingCallable</a>&nbsp;actor,
 <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html" title="class or interface in java.lang" class="external-link">Class</a>&lt;? extends <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html" title="class or interface in java.lang" class="external-link">Exception</a>&gt;&nbsp;clazz)</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab1 method-summary-table-tab4">
<div class="block">Use it to verify that an exception of specific type is thrown and to get
 access to the thrown exception (for further verifications).</div>
</div>
</div>
</div>
</div>
<div class="inherited-list">
<h3 id="methods-inherited-from-class-java.lang.Object">Methods inherited from class&nbsp;java.lang.<a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html" title="class or interface in java.lang" class="external-link">Object</a></h3>
<code><a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#clone()" title="class or interface in java.lang" class="external-link">clone</a>, <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#equals(java.lang.Object)" title="class or interface in java.lang" class="external-link">equals</a>, <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#finalize()" title="class or interface in java.lang" class="external-link">finalize</a>, <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#getClass()" title="class or interface in java.lang" class="external-link">getClass</a>, <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#hashCode()" title="class or interface in java.lang" class="external-link">hashCode</a>, <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#notify()" title="class or interface in java.lang" class="external-link">notify</a>, <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#notifyAll()" title="class or interface in java.lang" class="external-link">notifyAll</a>, <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#toString()" title="class or interface in java.lang" class="external-link">toString</a>, <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#wait()" title="class or interface in java.lang" class="external-link">wait</a>, <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#wait(long)" title="class or interface in java.lang" class="external-link">wait</a>, <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#wait(long,int)" title="class or interface in java.lang" class="external-link">wait</a></code></div>
</section>
</li>
</ul>
</section>
<section class="details">
<ul class="details-list">
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<li>
<section class="constructor-details" id="constructor-detail">
<h2>Constructor Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="&lt;init&gt;()">
<h3>CatchException</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="element-name"><a href="../../../src-html/com/googlecode/catchexception/CatchException.html#line-207">CatchException</a></span>()</div>
</section>
</li>
</ul>
</section>
</li>
<!-- ============ METHOD DETAIL ========== -->
<li>
<section class="method-details" id="method-detail">
<h2>Method Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="caughtException()">
<h3>caughtException</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="type-parameters">&lt;E extends <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html" title="class or interface in java.lang" class="external-link">Exception</a>&gt;</span>&nbsp;<span class="return-type">E</span>&nbsp;<span class="element-name"><a href="../../../src-html/com/googlecode/catchexception/CatchException.html#line-219">caughtException</a></span>()</div>
<div class="block">Returns the exception caught during the last call in the current thread.</div>
<dl class="notes">
<dt>Type Parameters:</dt>
<dd><code>E</code> - This type parameter makes some type casts redundant.</dd>
<dt>Returns:</dt>
<dd>Returns the exception caught during the last call in the current
 thread - if the call was made through a proxy that has been created via
 <a href="#verifyException(com.googlecode.catchexception.ThrowingCallable,java.lang.Class)"><code>verifyException()</code></a> or
 <a href="#catchException(com.googlecode.catchexception.ThrowingCallable,java.lang.Class)"><code>catchException()</code></a>. Returns
 null when no exception was caught.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="caughtException(java.lang.Class)">
<h3>caughtException</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="type-parameters">&lt;E extends <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html" title="class or interface in java.lang" class="external-link">Exception</a>&gt;</span>&nbsp;<span class="return-type">E</span>&nbsp;<span class="element-name"><a href="../../../src-html/com/googlecode/catchexception/CatchException.html#line-223">caughtException</a></span><wbr><span class="parameters">(<a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html" title="class or interface in java.lang" class="external-link">Class</a>&lt;E&gt;&nbsp;caughtExceptionType)</span></div>
</section>
</li>
<li>
<section class="detail" id="verifyException(com.googlecode.catchexception.ThrowingCallable)">
<h3>verifyException</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="element-name"><a href="../../../src-html/com/googlecode/catchexception/CatchException.html#line-244">verifyException</a></span><wbr><span class="parameters">(<a href="ThrowingCallable.html" title="interface in com.googlecode.catchexception">ThrowingCallable</a>&nbsp;actor)</span></div>
<div class="block">Use it to verify that an exception is thrown and to get access to the
 thrown exception (for further verifications).

 The following example verifies that obj.doX() throws a Exception:
 <code>verifyException(obj).doX(); // catch and verify
 assert "foobar".equals(caughtException().getMessage()); // further analysis
 </code>

 If <code>doX()</code> does not throw a <code>Exception</code>, then a
 <a href="ExceptionNotThrownAssertionError.html" title="class in com.googlecode.catchexception"><code>ExceptionNotThrownAssertionError</code></a> is thrown. Otherwise the thrown
 exception can be retrieved via <a href="#caughtException()"><code>caughtException()</code></a>.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>actor</code> - The instance that shall be proxied. Must not be
              <code>null</code>.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="verifyException(com.googlecode.catchexception.ThrowingCallable,java.lang.Class)">
<h3>verifyException</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="element-name"><a href="../../../src-html/com/googlecode/catchexception/CatchException.html#line-267">verifyException</a></span><wbr><span class="parameters">(<a href="ThrowingCallable.html" title="interface in com.googlecode.catchexception">ThrowingCallable</a>&nbsp;actor,
 <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html" title="class or interface in java.lang" class="external-link">Class</a>&lt;? extends <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html" title="class or interface in java.lang" class="external-link">Exception</a>&gt;&nbsp;clazz)</span></div>
<div class="block">Use it to verify that an exception of specific type is thrown and to get
 access to the thrown exception (for further verifications).

 The following example verifies that obj.doX() throws a MyException:
 <code>verifyException(obj, MyException.class).doX(); // catch and verify
 assert "foobar".equals(caughtException().getMessage()); // further analysis
 </code>

 If <code>doX()</code> does not throw a <code>MyException</code>, then a
 <a href="ExceptionNotThrownAssertionError.html" title="class in com.googlecode.catchexception"><code>ExceptionNotThrownAssertionError</code></a> is thrown. Otherwise the thrown
 exception can be retrieved via <a href="#caughtException()"><code>caughtException()</code></a>.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>actor</code> - The instance that shall be proxied. Must not be
              <code>null</code>.</dd>
<dd><code>clazz</code> - The type of the exception that shall be thrown by the
              underlying object. Must not be <code>null</code>.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="catchException(com.googlecode.catchexception.ThrowingCallable)">
<h3>catchException</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="element-name"><a href="../../../src-html/com/googlecode/catchexception/CatchException.html#line-290">catchException</a></span><wbr><span class="parameters">(<a href="ThrowingCallable.html" title="interface in com.googlecode.catchexception">ThrowingCallable</a>&nbsp;actor)</span></div>
<div class="block">Use it to catch an exception and to get access to the thrown exception
 (for further verifications).

 In the following example you catch exceptions that are thrown by
 obj.doX():
 <code>catchException(obj).doX(); // catch
 if (caughtException() != null) {
 assert "foobar".equals(caughtException().getMessage()); // further analysis
 }</code>
 If <code>doX()</code> throws a exception, then <a href="#caughtException()"><code>caughtException()</code></a>
 will return the caught exception. If <code>doX()</code> does not throw a
 exception, then <a href="#caughtException()"><code>caughtException()</code></a> will return <code>null</code>.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>actor</code> - The instance that shall be proxied. Must not be
              <code>null</code>.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="catchException(com.googlecode.catchexception.ThrowingCallable,java.lang.Class)">
<h3>catchException</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="element-name"><a href="../../../src-html/com/googlecode/catchexception/CatchException.html#line-319">catchException</a></span><wbr><span class="parameters">(<a href="ThrowingCallable.html" title="interface in com.googlecode.catchexception">ThrowingCallable</a>&nbsp;actor,
 <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html" title="class or interface in java.lang" class="external-link">Class</a>&lt;? extends <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html" title="class or interface in java.lang" class="external-link">Exception</a>&gt;&nbsp;clazz)</span></div>
<div class="block">Use it to catch an exception of a specific type and to get access to the
 thrown exception (for further verifications).

 In the following example you catch exceptions of type MyException that
 are thrown by obj.doX():
 <code>catchException(obj, MyException.class).doX(); // catch
 if (caughtException() != null) {
 assert "foobar".equals(caughtException().getMessage()); // further analysis
 }</code>
 If <code>doX()</code> throws a <code>MyException</code>, then
 <a href="#caughtException()"><code>caughtException()</code></a> will return the caught exception. If
 <code>doX()</code> does not throw a <code>MyException</code>, then
 <a href="#caughtException()"><code>caughtException()</code></a> will return <code>null</code>. If
 <code>doX()</code> throws an exception of another type, i.e. not a
 subclass but another class, then this exception is not thrown and
 <a href="#caughtException()"><code>caughtException()</code></a> will return <code>null</code>.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>actor</code> - The instance that shall be proxied. Must not be
              <code>null</code>.</dd>
<dd><code>clazz</code> - The type of the exception that shall be caught. Must not be
              <code>null</code>.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="catchException(com.googlecode.catchexception.ThrowingCallable,java.lang.Class,boolean)">
<h3>catchException</h3>
<div class="member-signature"><span class="modifiers">private static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="element-name"><a href="../../../src-html/com/googlecode/catchexception/CatchException.html#line-324">catchException</a></span><wbr><span class="parameters">(<a href="ThrowingCallable.html" title="interface in com.googlecode.catchexception">ThrowingCallable</a>&nbsp;actor,
 <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html" title="class or interface in java.lang" class="external-link">Class</a>&lt;? extends <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html" title="class or interface in java.lang" class="external-link">Exception</a>&gt;&nbsp;clazz,
 boolean&nbsp;assertException)</span></div>
</section>
</li>
<li>
<section class="detail" id="validateArguments(com.googlecode.catchexception.ThrowingCallable,java.lang.Class)">
<h3>validateArguments</h3>
<div class="member-signature"><span class="modifiers">private static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="element-name"><a href="../../../src-html/com/googlecode/catchexception/CatchException.html#line-347">validateArguments</a></span><wbr><span class="parameters">(<a href="ThrowingCallable.html" title="interface in com.googlecode.catchexception">ThrowingCallable</a>&nbsp;actor,
 <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html" title="class or interface in java.lang" class="external-link">Class</a>&lt;? extends <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html" title="class or interface in java.lang" class="external-link">Exception</a>&gt;&nbsp;clazz)</span></div>
</section>
</li>
<li>
<section class="detail" id="resetCaughtException()">
<h3>resetCaughtException</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="element-name"><a href="../../../src-html/com/googlecode/catchexception/CatchException.html#line-362">resetCaughtException</a></span>()</div>
<div class="block">Sets the <a href="#caughtException()"><code>caught exception</code></a> to null. This does
 not affect exceptions saved at threads other than the current one.

 Actually you probably never need to call this method because each method
 call on a proxied object in the current thread resets the caught
 exception. But if you want to improve test isolation or if you want to
 'clean up' after testing (to avoid memory leaks), call the method before
 or after testing.</div>
</section>
</li>
</ul>
</section>
</li>
</ul>
</section>
<!-- ========= END OF CLASS DATA ========= -->
</main>
<footer role="contentinfo">
<hr>
<p class="legal-copy"><small>Copyright &#169; 2011&#x2013;2022 <a href="https://github.com/hazendaz/">hazendaz</a>. All rights reserved.</small></p>
</footer>
</div>
</div>
</body>
</html>